proposals/0000-template-patterns.rst

```diff
@@ -0,0 +1,376 @@
+
+One notable extension of Miller's patterns is the extension by Duggan in his paper `"Unification with extended patterns" <https://doi.org/10.1016/S0304-3975(97)00141-2>`_. Duggan extends the simple template application form to allow projections of local binders as arguments to the template variables. Such an extension might also be applicable to the template application form in this proposal.
+
+Examples
+--------
+
+* 	One of the simplest examples is this rule:
+	::
+
+		{-# RULES "foo" forall f. foo (\x -> f x) = "RULE FIRED" #-}
+
+	It would match expressions like:
+	::
+
+		foo (\x -> x * 2 + x)
+
+* 	The higher order pattern may involve multiple locally bound variables, e.g.:
+	::
+
+		{-# RULES "foo" forall f. foo (\x y z -> f x y z) = "RULE FIRED" #-}
+
+	Which would match:
+	::
+
+		foo (\x y z -> x * y + z)
+
+	But not every variable has to occur in the match. It would also match this expression where ``y`` does not occur:
+	::
+
+		foo (\x y z -> x * 2 + z)
+
+* 	Locally bound variables may only occur once.
+	Consider the following rule:
+	::
+
+		{-# RULES "foo" forall f. foo (\x -> f x x) = "RULE FIRED" #-}
+
+	This would **not** match:
+	::
+
+		foo (\x -> x * 2 + x)
+
+	But it does contain the valid subrule ``f x``, so it would match:
+	::
+
+		foo (\x -> (bar x . baz) x)
+
+* 	Similarly if the template variable ``f`` is applied to non-variable arguments then it only matches a literal application.
+	Consider this rule:
+	::
+
+		{-# RULES "foo" forall f. foo (\x y -> f x 2 y) = "RULE FIRED" #-}
+
+	This would **not** match:
+	::
+
+		foo (\x y -> x * 2 + y)`
+
+	But again it does contain the higher order pattern ``f x``, so it would match:
+	::
+
+		foo (\x y -> (bar x . baz) 2 y)
+
+Effect and Interactions
+-----------------------
+
+The main effect of this proposal is that rewrite rules involving higher order patterns now match more expressions.
+But the additional matches are guaranteed to be beta equivalent, so this change does not cause existing rules to become semantically incorrect.
+
+The only contentious interactions could occur due to rules that now overlap under the new rules, for example:
+::
+
+	{-# RULES
+	"foo->bar"  forall f x.  foo x (\y. f y) = bar x f
+	"foo->baz"  forall   x.  foo x (\y. y * 2 + y) = baz x
+	#-}
+
+Previously, the rule ``"foo->baz"`` would always fire when encountering the expression ``foo x (\y. y * 2 + y)``, but now the rule ``"foo->bar"`` also matches.
+However, we do not expect that this occurs in practice.
```

Comment:
Well, except when we implement our new `mapList` RULE that will overlap with every FB rule, no?

I think we should simply let "foo->baz" win. It's strictly more specific than "foo->bar", plus that's what backwards compatibility demands.

Reply:
Good point.  And it will, already. I'll add words